### ISSUE TITLE
loss of training swav explode after introducing queue

### ISSUE THREAD

**Issue by hanxx862** [lightly-ai/lightly#1148]
Hi Dear, 

I am using the Swav queue method and code shown as below

```python

class SwaV(pl.LightningModule):
    def __init__(self):
        super().__init__()
        resnet = torchvision.models.resnet50()
        self.backbone = nn.Sequential(*list(resnet.children())[:-1])
        self.projection_head = SwaVProjectionHead(2048, 2048, 128)
        self.prototypes = SwaVPrototypes(128, 1000, 750)
        self.start_queue_at_epoch = 10
        self.queues = nn.ModuleList([MemoryBankModule(size=1024) for _ in range(2)])
        self.criterion = SwaVLoss(sinkhorn_gather_distributed=True, sinkhorn_epsilon = 0.03)
    
    def forward(self, x):
        x = self._subforward(x)
        return self.prototypes(x)

    def training_step(self, batch, batch_idx):
        batch_swav, _, _ = batch
        high_resolution, low_resolution = batch_swav[:2], batch_swav[2:]
        self.prototypes.normalize()

        high_resolution_features = [self._subforward(x) for x in high_resolution]
        low_resolution_features = [self._subforward(x) for x in low_resolution]

        high_resolution_prototypes = [
            self.prototypes(x, self.global_step) for x in high_resolution_features
        ]
        low_resolution_prototypes = [
            self.prototypes(x, self.global_step) for x in low_resolution_features
        ]
        queue_prototypes = self._get_queue_prototypes(high_resolution_features)
        loss = self.criterion(
            high_resolution_prototypes, low_resolution_prototypes, queue_prototypes
        )
        self.log("loss", loss,  on_epoch=True, logger=True, batch_size=batch_size,  sync_dist=True)
        cur_lr = self.trainer.optimizers[0].param_groups[0]['lr']
        self.log("lr", cur_lr, prog_bar=True)
        return loss
    # see the https://github.com/lightly-ai/lightly/blob/master/lightly/models/modules/heads.py#L449
    # that parameter in self.prototypes should be step rather than epoch...
    
    
    def validation_step(self, batch, batch_idx):
        batch_swav, _, _ = batch
        high_resolution, low_resolution = batch_swav[:2], batch_swav[2:]
    
        self.prototypes.normalize()
        high_resolution_features = [self._subforward(x) for x in high_resolution]
        low_resolution_features = [self._subforward(x) for x in low_resolution]

        high_resolution_prototypes = [
                self.prototypes(x, self.global_step) for x in high_resolution_features
            ]
        low_resolution_prototypes = [
                self.prototypes(x, self.global_step) for x in low_resolution_features
            ]
        queue_prototypes = self._get_queue_prototypes(high_resolution_features)
        loss = self.criterion(
                high_resolution_prototypes, low_resolution_prototypes, queue_prototypes
            )
        self.log("val_loss", loss,  on_epoch=True, logger=True, batch_size=batch_size,  sync_dist=True)
        return loss


    def configure_optimizers(self):
        #optim = torch.optim.Adam(self.parameters(), lr=0.0001)
        #optim = LARS(optim, )
        # torch.optim.lr_scheduler.ReduceLROnPlateau(optim)
        optim = LARS(self.parameters(), lr = 0.6, momentum = 0.9, weight_decay = 1e-6)
        #optim = torch.optim.Adam(
            #self.parameters(),
            #lr=1e-3 * 0.25,
            #weight_decay=1e-6,
        #)
        lr_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optim, T_max = 200 * 720, eta_min = 0.0006)
        return  {
            "optimizer": optim,
            "lr_scheduler": {
                "scheduler" : lr_scheduler,
                "interval": "step",
                'frequency': 1
            }
        }
            

    def _subforward(self, input):
        features = self.backbone(input).flatten(start_dim=1)
        features = self.projection_head(features)
        features = nn.functional.normalize(features, dim=1, p=2)
        return features
    

    @torch.no_grad()
    def _get_queue_prototypes(self, high_resolution_features):
        if len(high_resolution_features) != len(self.queues):
            raise ValueError(
                f"The number of queues ({len(self.queues)}) should be equal to the number of high "
                f"resolution inputs ({len(high_resolution_features)}). Set `n_queues` accordingly."
            )

        # Get the queue features
        queue_features = []
        for i in range(len(self.queues)):
            _, features = self.queues[i](high_resolution_features[i], update=True)
            # Queue features are in (num_ftrs X queue_length) shape, while the high res
            # features are in (batch_size X num_ftrs). Swap the axes for interoperability.
            features = torch.permute(features, (1, 0))
            queue_features.append(features)

        # If loss calculation with queue prototypes starts at a later epoch,
        # just queue the features and return None instead of queue prototypes.
        if (
            self.start_queue_at_epoch > 0
            and self.current_epoch < self.start_queue_at_epoch
        ):
            return None

        # Assign prototypes
        queue_prototypes = [
            self.prototypes(x, self.global_step) for x in queue_features
        ]
        return queue_prototypes
```

After 30 epoch, the queue will be introduced and my trainings loss jump from 4.0x to 6.1x,  did I make any mistake in the model? I am trying to decay the learning rate based on the step rather than epoch.  Thanks!

**Comment by hanxx862** [comment]
The training result on the same dataset is really different from original implementation swav in https://github.com/facebookresearch/swav/blob/main/main_swav.py,  after 30 epoch.  Thanks!

**Comment by hanxx862** [comment]
@ibro45

**Comment by hanxx862** [comment]
Any suggestion will be appreciated

**Comment by guarin** [comment]
Hi! Thanks for using Lightly!

Looking over your code I cannot see any obvious issues. Could you give some more info on the dataset and training (dataset, batch size, number of images, num gpus, ..)? You mentioned that you compared against the SwaV implementation in vissl and that the training result differs. What exactly do you mean by the training result? Just the accuracy or also the loss?

**Comment by guarin** [comment]
Closing due to inactivity
